### PR TITLE
Add maximum number of retries to PingUploadWorker

### PR DESCRIPTION
**Pull request by brizental**
Fixes [Bug 1644290](https://bugzilla.mozilla.org/show_bug.cgi?id=1644290)

This is a temporary fix so we can stop [this issue](https://github.com/mozilla-mobile/fenix/issues/11335). 

I opened a bug to move this retrying logic to the Rust side: https://bugzilla.mozilla.org/show_bug.cgi?id=1644364

### CODE REVIEW

File: glean-core/android/src/main/java/mozilla/telemetry/glean/scheduler/PingUploadWorker.kt [hunk 1]
```diff
@@ -112,14 +121,20 @@ class PingUploadWorker(context: Context, params: WorkerParameters) : Worker(cont
...
                     // Process the upload response
-                    LibGleanFFI.INSTANCE.glean_process_ping_upload_response(incomingTask, result.toFfi())
+                    LibGleanFFI.INSTANCE.glean_process_ping_upload_response(incomingTask, result)
                 }
                 PingUploadTask.Wait -> return Result.retry()
                 PingUploadTask.Done -> return Result.success()
             }
-        } while (true)
+        } while (uploadFailures < MAX_RETRIES)
+        return Result.failure()
```
Comment (brizental) [inline glean-core/android/src/main/java/mozilla/telemetry/glean/scheduler/PingUploadWorker.kt:138]: Unsure if this should really be a `failure` or a `retry`.

Comment (badboy) [reply glean-core/android/src/main/java/mozilla/telemetry/glean/scheduler/PingUploadWorker.kt:138]: We use a `OneTimeWorkRequest` which defaults to exponential backoff on `retry`.
I think `failure` is the right thing here. If we fail too often, we don't want to reschedule it at all. When a ping gets submitted again (e.g. next time the app is backgrounded) any pending pings are picked up again anyway.

Comment (badboy) [reply glean-core/android/src/main/java/mozilla/telemetry/glean/scheduler/PingUploadWorker.kt:138]: cc @Dexterp37 just for a quick look

File: glean-core/android/src/main/java/mozilla/telemetry/glean/scheduler/PingUploadWorker.kt [hunk 2]
```diff
@@ -112,14 +121,20 @@ class PingUploadWorker(context: Context, params: WorkerParameters) : Worker(cont
                         action.request.body,
                         action.request.headers,
                         Glean.configuration
-                    )
+                    ).toFfi()
+
+                    if (result == Constants.UPLOAD_RESULT_RECOVERABLE ||
+                        result == Constants.UPLOAD_RESULT_UNRECOVERABLE) {
```
Comment (badboy) [inline glean-core/android/src/main/java/mozilla/telemetry/glean/scheduler/PingUploadWorker.kt:127]: A thing I only notice now:

* We know that `unrecoverable` failures will lead to the file being deleted, we would only need to count the recoverable errors here.
  * HTTP status 400-499 are also unrecoverable, at least internally to the Rust part. Again, we don't need to care, the ping is deleted and not retried.
* android-components never uses `UPLOAD_RESULT_UNRECOVERABLE` anyway (there we catch all I/O errors as `recoverable`

If we do fail on I/O in Fenix we correctly catch that here with the `recoverable` part.

**Comment by mdboom** [comment]
Should we file a follow up bug for Swift and Python?

**Comment by brizental** [comment]
> Should we file a follow up bug for Swift and Python?

https://bugzilla.mozilla.org/show_bug.cgi?id=1644419

File: glean-core/android/src/main/java/mozilla/telemetry/glean/scheduler/PingUploadWorker.kt (hunk 1 above)
Comment (Dexterp37) [reply glean-core/android/src/main/java/mozilla/telemetry/glean/scheduler/PingUploadWorker.kt:137]: Works for me, assuming failure doesn't mean deleting the pings

Comment (badboy) [reply glean-core/android/src/main/java/mozilla/telemetry/glean/scheduler/PingUploadWorker.kt:137]: Nope, it's only marking the worker as failed. The pings are handled inside glean-core and will be retried by the next worker.

**Pull request merged**